uefi-logger/src/writer.rs

```diff
@@ -37,27 +38,33 @@ impl fmt::Write for OutputWriter {
             let mut add_char = |ch| {
                 // UEFI only supports UCS-2 characters, not UTF-16,
                 // so there are no multibyte characters.
-                let ch = ch as u16;
-
                 buf[i] = ch;
-
                 i += 1;
 
                 if i == BUF_SIZE {
-                    flush_buffer(&mut buf, &mut i)
+                    match flush_buffer(&mut buf, &mut i) {
+                        Ok(()) => { Ok(()) },
+                        Err(_) => { Err(Status::ProtocolError) },
+                    }
                 } else {
                     Ok(())
                 }
             };
-
-            for ch in s.chars() {
-                if ch == '\n' {
-                    // Prepend an '\r'.
-                    add_char('\r')?;
+            let mut add_ch = |ch| {
```

Comment:
This could be refactored as such:

```rust
let mut add_ch = |ch| {
  add_char(ch)?;

  if ch == '\n' as u16 {
    add_char('\r' as u16)
  } else {
    Ok(())
  }
};
```

That way you don't need to explicitly `match` the `Result`.